diff --git a/src/include/lwip/sys.h b/src/include/lwip/sys.h
index 168e465..e03e164 100644
--- a/src/include/lwip/sys.h
+++ b/src/include/lwip/sys.h
@@ -443,7 +443,7 @@ u32_t sys_jiffies(void);
  * Not implementing this function means you cannot use some modules (e.g. TCP
  * timestamps, internal timeouts for NO_SYS==1).
  */
-u32_t sys_now(void);
+//u32_t sys_now(void);
 
 /* Critical Region Protection */
 /* These functions must be implemented in the sys_arch.c file.
